apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Trace/TraceTest.cs

```diff
@@ -0,0 +1,420 @@
+            var trace = await GetTrace(spanName, startTime);
+
+            Assert.NotNull(trace);
+            Assert.Equal(2, trace.Spans.Count);
+            var span = trace.Spans.First(s => s.Name.StartsWith("Trace"));
+            Assert.Single(span.Labels);
+            Assert.Contains(nameof(TraceController), span.Labels[Common.Labels.StackTrace]);
+            Assert.Contains(nameof(TraceController.CreateStackTrace), span.Labels[Common.Labels.StackTrace]);   
+        }
+
+        [Fact]
+        public async Task Trace_Header()
+        {
+            string traceId = TraceIdFactory.Create().NextId();
+            ulong spanId = SpanIdFactory.Create().NextId();
+
+            string testId = Utils.GetTestId();           
+            var startTime = Timestamp.FromDateTime(DateTime.UtcNow);
+            var builder = new WebHostBuilder().UseStartup<TraceTestNoBufferLowQpsApplication>();
+            using (TestServer server = new TestServer(builder))
+            {
+                var client = server.CreateClient();
+                
+                var header = TraceHeaderContext.Create(traceId, spanId, true);
+                client.DefaultRequestHeaders.Add(TraceHeaderContext.TraceHeader, header.ToString());
+                await client.GetAsync($"/Trace/Trace/{testId}");
+            }
+
+            var spanName = TraceController.GetMessage(nameof(TraceController.Trace), testId);
+            var trace = await GetTrace(spanName, startTime);
+
+            Assert.NotNull(trace);
+            Assert.Equal(traceId, trace.TraceId);
+            Assert.Equal(2, trace.Spans.Count);
+            var span = trace.Spans.First(s => s.Name.StartsWith("/Trace"));
+            Assert.Equal(spanId, span.ParentSpanId);
+        }
+
+        [Fact]
+        public async Task Trace_QPS()
+        {
+            string testId = Utils.GetTestId();
+            var startTime = Timestamp.FromDateTime(DateTime.UtcNow);
+
+            var builder = new WebHostBuilder().UseStartup<TraceTestNoBufferApplication>();
+            using (TestServer server = new TestServer(builder))
+            {
+                var client = server.CreateClient();
+
+                // The first request is not traced, sleep long enough 
+                // for the rate limiter to allow a request, then the second
+                // request is traced.
+                await client.GetAsync($"/Trace/TraceLabels/{testId}");
+                Thread.Sleep(TimeSpan.FromSeconds(2));
+                await client.GetAsync($"/Trace/Trace/{testId}");                
+            }
+
+            var spanNameTrace = TraceController.GetMessage(nameof(TraceController.Trace), testId);
+            var spanNameLabels = TraceController.GetMessage(nameof(TraceController.TraceLabels), testId);
+            Assert.Null(await GetTrace(spanNameLabels, startTime, expectTrace: false));
+            Assert.NotNull(await GetTrace(spanNameTrace, startTime));
```

Comment:
I agree that this polling is subpar.  It's go generous with sleeps due to API quota issues.  The tests are only run before a release so it's not being run often.

Either way I do have an action item to make a better common polling solution for tests.